cpp/src/text/jaccard.cu

```diff
@@ -0,0 +1,305 @@
+ * and |x| is the number of unique values in x.
+ */
+struct jaccard_fn {
+  cudf::size_type const* d_uniques1;
+  cudf::size_type const* d_uniques2;
+  cudf::size_type const* d_intersects;
+
+  __device__ float operator()(cudf::size_type idx) const
+  {
+    auto const count1     = d_uniques1[idx];
+    auto const count2     = d_uniques2[idx];
+    auto const intersects = d_intersects[idx];
+    // the intersect values are in both sets so a union count
+    // would need to subtract the intersect count from one set
+    // (see formula in comment above)
+    auto const unions = count1 + count2 - intersects;
+    return unions ? ((float)intersects / (float)unions) : 0.f;
+  }
+};
+
+/**
+ * @brief Create hashes for each substring
+ *
+ * Uses the hash_character_ngrams to hash substrings of the input column
+ * which returns a lists column where each row is the hashes for the substrings
+ * of the corresponding input string row.
+ *
+ * The hashes are then sorted using a segmented-sort as setup to
+ * perform the unique and intersect operations.
+ */
+std::unique_ptr<cudf::column> hash_substrings(cudf::strings_column_view const& col,
+                                              cudf::size_type width,
+                                              rmm::cuda_stream_view stream)
+{
+  auto hashes = hash_character_ngrams(col, width, stream, rmm::mr::get_current_device_resource());
+  auto const input   = cudf::lists_column_view(hashes->view());
+  auto const offsets = input.offsets_begin();
+  auto const data    = input.child().data<uint32_t>();
+
+  rmm::device_uvector<uint32_t> sorted(input.child().size(), stream);
+
+  // this is wicked fast and much faster than using cudf::lists::detail::sort_list
+  rmm::device_buffer d_temp_storage;
+  size_t temp_storage_bytes = 0;
+  cub::DeviceSegmentedSort::SortKeys(d_temp_storage.data(),
+                                     temp_storage_bytes,
+                                     data,
+                                     sorted.data(),
+                                     sorted.size(),
+                                     input.size(),
+                                     offsets,
+                                     offsets + 1,
+                                     stream.value());
+  d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream};
+  cub::DeviceSegmentedSort::SortKeys(d_temp_storage.data(),
+                                     temp_storage_bytes,
+                                     data,
+                                     sorted.data(),
+                                     sorted.size(),
+                                     input.size(),
+                                     offsets,
+                                     offsets + 1,
+                                     stream.value());
+
+  auto contents = hashes->release();
+  return cudf::make_lists_column(
+    col.size(),
+    std::move(contents.children.front()),
```

Comment:
Is this taking advantage of the fact the hashes has the same offsets as the input column so you can avoid allocating a new set of offsets? If so, that would be helpful to add a comment for since it is confusing to see a part of the hashes going into the output.

Reply:
The offsets are the same because the segmented-sort does not change them.